File: zennote/tabview.py
from gi.repository import (
    Adw,
    GObject,
    Gdk,
    Gio,
    Gtk,
)  # pyright: ignore[reportMissingModuleSource]
from zennote.dialogs import UnsavedResponse, request_open_file, unsaved_dialog
from zennote.editor import Editor

import pathlib
from typing import cast

from zennote.utils import Args, KwArgs


@Gtk.Template(resource_path="/ui/tabview.ui")
# kann nicht von Adw.ToolbarView verbt werden maaaan
class EditorTabView(Adw.Bin):
    __gtype_name__: str = "EditorTabView"
    bar: Adw.TabBar = cast(Adw.TabBar, Gtk.Template.Child("tab-bar"))
    view: Adw.TabView = cast(Adw.TabView, Gtk.Template.Child("tab-view"))

    opened_files: dict[pathlib.Path, tuple[Adw.TabPage, Editor]] = {}

    def __init__(self, **_kwargs: KwArgs):
        super().__init__()
        # Connect the tab view to the tab bar
        self.bar.set_view(self.view)
        self.new_file()

    def open_file(self, path: pathlib.Path):
        """
        opens a path to a file in the tabview.
        path may not be a valid file, in which case it will simply not open
        """

        editor = Editor()
        editor.open_file(path)
        page = self.view.append(editor)

        name = pathlib.Path(path).name
        page.set_title(name)
        self._setup_editor_bindings(page, editor)
        self.view.set_selected_page(page)

    def open_file_with_dialog(self):
        def on_open(f: Gio.File | None):
            if not f:
                return
            path = f.get_path()
            if path:
                self.open_file(pathlib.Path(path))

        request_open_file(on_open)

    def get_active_editor(self) -> Editor | None:
        page = self.view.get_selected_page()

        if not page:
            return None

        return cast(Editor, page.get_child())

    def _setup_editor_bindings(self, page: Adw.TabPage, editor: Editor):
        editor.bind_property(
            "filename",
            page,
            "title",
            GObject.BindingFlags.BIDIRECTIONAL | GObject.BindingFlags.SYNC_CREATE,
        )

    @Gtk.Template.Callback()
    def _on_close(
        self, view: Adw.TabView, page: Adw.TabPage, *_args: Args, **_kwargs: KwArgs
    ):
        editor: Editor = cast(Editor, page.get_child())

        if editor.is_saved():
            view.close_page_finish(page, True)
            return Gdk.EVENT_STOP

        def cb(res: UnsavedResponse):
            match res:
                case "save":

                    def cb(close: bool):
                        view.close_page_finish(page, close)

                    editor.write_to_file(cb)

                case "cancel":
                    view.close_page_finish(page, False)
                case "discard":
                    view.close_page_finish(page, True)

        print("file unsaved")
        unsaved_dialog(cb, editor.get_filename())

        return Gdk.EVENT_STOP

    # Für Kontextmenü - Todo
    @Gtk.Template.Callback()  # pyright: ignore[reportAny,]
    def _on_context(self, *_args: Args, **_kwargs: KwArgs):
        print(_args)
        print(_kwargs)

    def new_file(self) -> Editor:
        editor = Editor(saved=True)

        page = self.view.prepend(editor)

        self._setup_editor_bindings(page, editor)

        self.view.set_selected_page(page)

        return editor

    def close_active(self):
        page = self.view.get_selected_page()

        if page:
            self.view.close_page(page)

File: src/tabview.rs
use adw::subclass::prelude::ObjectSubclassIsExt;
use gtk::gio;
use gtk::glib::object::{CastNone, ObjectExt};
use gtk::glib::{self, derived_properties, Object};
use gtk::prelude::{GtkWindowExt, WidgetExt};
use gtk::subclass::prelude::{ObjectImpl, ObjectSubclass, ObjectSubclassExt};
use gtk::template_callbacks;
use gtk::CompositeTemplate;
use std::cell::{Ref, RefCell};
use std::io;
use std::path::PathBuf;
use std::time::Duration;

use crate::editor::Editor;
use crate::utils;

mod inner {

    use std::{cell::Cell, env, path::PathBuf};

    use adw::{
        subclass::{bin::BinImpl, prelude::ObjectImplExt},
        TabView,
    };
    use gtk::{
        glib::object::Cast,
        subclass::{prelude::*, widget::WidgetImpl, window::WindowImpl},
    };

    use crate::{editor, tabview, utils::UnsavedResponses};

    use super::*;
    #[derive(CompositeTemplate, Default)]
    #[template(resource = "/io/github/bracket/tabview.ui")]
    pub struct EditorTabView {
        #[template_child(id = "tabview")]
        pub tabview: TemplateChild<adw::TabView>,
        #[template_child(id = "tabbar")]
        pub tabbar: TemplateChild<adw::TabBar>,
    }
    #[template_callbacks]
    impl EditorTabView {
        #[template_callback]
        fn on_close(&self, page: &adw::TabPage, tabview: &adw::TabView) -> bool {
            let Ok(editor) = page.child().downcast::<Editor>() else {
                tabview.close_page_finish(page, false);
                return true;
            };

            if editor.saved() {
                println!("saved");
                tabview.close_page_finish(page, true);
                return true;
            }

            {
                let tabview = tabview.clone();
                let page = page.clone();
                let fut = async move {
                    let response = utils::prompt_unsaved_changes(None, None)
                        .await
                        .expect("failed to parse response of unsaved dialog.");

                    match response {
                        UnsavedResponses::Cancel => {
                            tabview.close_page_finish(&page, false);
                        },
                        UnsavedResponses::Discard => {
                            tabview.close_page_finish(&page, true);
                        },
                        UnsavedResponses::Save => {
                            if let Some(editor) = page.child().downcast_ref::<Editor>() {
                                editor.save_file_fut().await;
                            }
                            tabview.close_page_finish(&page, true);
                        }
                    }
                };
                glib::MainContext::default().spawn_local(fut);
            }


            return true;
        }
    }
    #[glib::object_subclass]
    impl ObjectSubclass for EditorTabView {
        const NAME: &'static str = "EditorTabView";
        type Type = super::EditorTabView;
        type ParentType = adw::Bin;

        fn new() -> Self {
            Self {
                ..Default::default()
            }
        }
        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
            klass.bind_template_callbacks();
        }
        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }
    impl ObjectImpl for EditorTabView {
        fn constructed(&self) {
            self.parent_constructed();
            let obj = self.obj();

            let imp = obj.imp();
            imp.tabbar.set_view(Some(&imp.tabview));
            obj.new_file(true);
        }
    }
    impl WidgetImpl for EditorTabView {}
    impl BinImpl for EditorTabView {}
}

glib::wrapper! {
    pub struct EditorTabView(ObjectSubclass<inner::EditorTabView>)
    @extends adw::Bin, gtk::Widget,
    @implements gtk::Accessible, gtk::Buildable, gtk::ConstraintTarget;
}

impl EditorTabView {
    pub fn new() -> Self {
        let obj = Object::new();
        // inner::EditorTabView::from_obj(&obj)
        //     .working_dir
        //     .replace(Some(dir));
        obj
    }
    fn set_up_editor_bindings(&self, page: &adw::TabPage, editor: &Editor) {
        editor
            .bind_property("filename", page, "title")
            .bidirectional()
            .sync_create()
            .build();
    }
    pub fn new_file(&self, focus: bool) {
        let tabview = &self.imp().tabview;
        let editor = Editor::new();

        let page = tabview.prepend(&editor);

        self.set_up_editor_bindings(&page, &editor);

        if focus {
            tabview.set_selected_page(&page);
        }
    }
    pub fn open_file(&self, path: &PathBuf, focus: bool) -> Result<(), io::Error> {
        let tabview = &self.imp().tabview;
        let editor = Editor::new();

        editor.open_file_path(path)?;

        let page = tabview.prepend(&editor);

        self.set_up_editor_bindings(&page, &editor);

        tabview.set_selected_page(&page);

        if focus {
            tabview.set_selected_page(&page);
        }
        Ok(())
    }
    pub fn get_active_editor(&self) -> Option<Editor> {
        self.imp()
            .tabview
            .selected_page()
            .map(|p| p.child())
            .and_downcast::<Editor>()
    }
    pub fn close_active(&self) {
        let tabview = &self.imp().tabview;

        if let Some(selected) = tabview.selected_page() {
            tabview.close_page(&selected);
        }
    }
}
